test/utils/ktesting/tcontext.go

```diff
@@ -0,0 +1,195 @@
+limitations under the License.
+*/
+
+package ktesting
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"testing"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	_ "k8s.io/component-base/logs/testinit"
+	"k8s.io/klog/v2"
+	"k8s.io/klog/v2/ktesting"
+	"k8s.io/kubernetes/test/utils/format"
+)
+
+// TContext combines [context.Context], [testing.TB] and some additional
+// methods.  Log output is associated with the current test. Errors ([Error],
+// [Errorf]) are recorded with "ERROR" as prefix, fatal errors ([Fatal],
+// [Fatalf]) with "FATAL ERROR".
+type TContext interface {
+	context.Context
+	testing.TB
+
+	// Cancel can be invoked to cancel the context before the test is completed.
+	// Tests which use the context to control goroutines and then wait for
+	// termination of those goroutines must call Cancel to avoid a deadlock.
+	Cancel()
+
+	// Logger returns a logger for the current test.
+	Logger() klog.Logger
+
+	// TB returns the underlying testing.TB. This can be used to "break
+	// the glass" and cast back into a testing.T or testing.TB.
+	TB() testing.TB
+
+	Client() clientset.Interface
+	Dynamic() dynamic.Interface
+	APIExtensions() apiextensions.Interface
+}
+
+// Init can be called in a unit or integration test to create
+// a test context which:
+// - has a per-test logger with verbosity derived from the -v command line flag
+// - gets canceled when the test finishes (via tl.Cleanup)
+//
+// Note that the test context supports the interfaces of testing.TB and
+// context.Context and thus can be used like one of those where needed.
+// It also has additional methods for retrieving the logger and canceling
+// the context early, which can be useful in tests which want to wait
+// for goroutines to terminate after cancellation.
+//
+// Can be called more than once per test to get different contexts with
+// independent cancelation.
+func Init(tb testing.TB) TContext {
+	config := ktesting.NewConfig(
+		ktesting.AnyToString(func(v interface{}) string {
+			return format.Object(v, 1)
+		}),
+		ktesting.VerbosityFlagName("v"),
+		ktesting.VModuleFlagName("vmodule"),
+	)
+
+	// Copy klog settings instead of making the ktesting logger
+	// configurable directly.
+	var fs flag.FlagSet
+	config.AddFlags(&fs)
+	for _, name := range []string{"v", "vmodule"} {
+		from := flag.CommandLine.Lookup(name)
+		to := fs.Lookup(name)
+		if err := to.Value.Set(from.Value.String()); err != nil {
+			panic(err)
+		}
+	}
+
+	logger := ktesting.NewLogger(tb, config)
+	ctx := klog.NewContext(context.Background(), logger)
+	return WithCancel(ctx, tb)
+}
+
+// WithCancel sets up cancelation in a [tb.Cleanup] and
+// constructs a new TContext. In contrast to [Init],
+// it uses the logger which is already available via
+// the context parameter or the global fallback logger.
+func WithCancel(ctx context.Context, tb testing.TB) TContext {
```

Comment:
`WithCancel` was needed in one place where `ktesting.Init` seemed less suitable. I'm now leaning more towards adding `WithCancel` and `WithTimeout` methods to `TContext`. I'll explore that...

Reply:
I implemented several `ktesting.With*` functions. `scheduler_perf_test.go` shows how they are used. I like this better than the previous ad-hoc solution.